src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,205 @@
+       for (int j = 0; j < numFun; j++)
+       fitness += function.Evaluate(parameters, j);
+
+       // Save fitness values
+       fitnessValues[i] = fitness;
+       fitness = 0;
+     }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+      << (double)fitnessValues.min() << std::endl;
+
+      // create the next generation of species
+      Reproduce();
+
+      // check for termination
+      if (finalValue >= (double)fitnessValues.min())
+        {
+          Log::Info << "Terminating. Given fitness criteria " << finalValue
+          << " > " << (double)fitnessValues.min() << std::endl;
+          break;
+        }
+
+      // check for termination
+      if (fitnessHist != DBL_MAX && gen != 1 &&
+        (lastBestFitness - (double)fitnessValues.min()) < fitnessHist)
+        {
+          Log::Info << "Terminating. Fitness History change "
+          << (lastBestFitness - (double)fitnessValues.min())
+          << " < " << fitnessHist << std::endl;
+          break;
+        }
+
+      // Store the best fitness of this generation before update
+      lastBestFitness = (double)fitnessValues.min();
+  }
+
+  // Set the best candidate into the network
+  answer = population.submat(index[0], 0, index[0], populationSize-1);
+
+  // find the best fitness
+  for (int j = 0; j < numFun; j++)
+      fitness += function.Evaluate(answer, j);
+
+  return fitness;
+}
+
+//! Reproduce candidates to create the next generation
+void CNE::Reproduce()
+{
+  // Sort fitness value. The smaller the better
+  index = arma::sort_index(fitnessValues);
+
+  // Find the number of elite percentage
+  size_t numElite = floor(selectPercent * populationSize);
+
+  // Making sure we have even number of candidates to remove and create
+  if ((populationSize - numElite) % 2 != 0) numElite++;
+
+  for (size_t i = numElite; i < populationSize-1; i++)
+  {
+    // Select 2 parents from the elite group randomly [0, numElite)
+    size_t mom = mlpack::math::RandInt(0, numElite);
+    size_t dad = mlpack::math::RandInt(0, numElite);
+
+    // Crossover parents to create 2 childs replacing the droped out candidates
+    Crossover(mom, dad, i, i+1);
+  }
+
+  // Mutate the weights with small noise.
+  // This is done to bring change in the next generation.
+  Mutate();
+}
+
+//! Crossover parents to create new childs
+void CNE::Crossover(size_t mom, size_t dad, size_t child1, size_t child2)
+{
+  // find the index of these candidates in the population matrix
+  mom = index[mom];
+  dad = index[dad];
+  child1 = index[child1];
+  child2 = index[child2];
+
+  // Remove the cadidates and instead place the parents
+  population.row(child1) = population.row(mom);
+  population.row(child2) = population.row(dad);
+
+  double rand;
+  // Randomly alter mom and dad genome data to get two childs
+  for (size_t i = 0; i < population.n_cols; i++)
+  {
+    // Select a random value from the normal distribution
+    rand = mlpack::math::RandNormal();
```

Comment:
Let us use a uniform random number between 0 and 1 (``mlpack::math::Random``) here and ``0.5`` as condition.